src/NationalInstruments.Analyzers/Correctness/ApprovedNamespaceAnalyzer.cs

```diff
@@ -0,0 +1,252 @@
+        /// </summary>
+        public static DiagnosticDescriptor TestRule { get; } = new DiagnosticDescriptor(
+            DiagnosticId,
+            new LocalizableResourceString(nameof(Resources.NI1800_TestTitle), Resources.ResourceManager, typeof(Resources)),
+            new LocalizableResourceString(nameof(Resources.NI1800_TestMessage), Resources.ResourceManager, typeof(Resources)),
+            Resources.CategoryNamespaces,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: new LocalizableResourceString(nameof(Resources.NI1800_TestDescription), Resources.ResourceManager, typeof(Resources)),
+            helpLinkUri: string.Empty);
+
+        /// <summary>
+        /// Rule for failure to read approved namespaces files.
+        /// </summary>
+        public static DiagnosticDescriptor FileReadRule { get; } = new DiagnosticDescriptor(
+            $"{DiagnosticId}_ReadError",
+            new LocalizableResourceString(nameof(Resources.NI1800_FileReadErrorTitle), Resources.ResourceManager, typeof(Resources)),
+            new LocalizableResourceString(nameof(Resources.NI1800_FileReadErrorMessage), Resources.ResourceManager, typeof(Resources)),
+            Resources.CategoryNamespaces,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        /// <inheritdoc/>
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(ProductionRule, TestRule, FileReadRule);
+
+        /// <inheritdoc/>
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecutionIf(IsRunningInProduction && !InDebugMode);
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        /// <summary>
+        /// Adds a namespace to the approved namespaces list
+        /// </summary>
+        /// <param name="namespaceName">Namespace to be approved</param>
+        internal static void ApproveNamespace(string namespaceName)
+        {
+            var namespacesFilePath = IsTestNamespace(namespaceName) ? _approvedTestNamespacesFilePath : _approvedNamespacesFilePath;
+            var lines = File.ReadAllLines(namespacesFilePath);
+            var namespaces = lines
+                    .Concat(new[] { namespaceName })
+                    .Select(x => x.Trim())
+                    .OrderBy(x => x)
+                    .Distinct();
+            File.WriteAllLines(namespacesFilePath, namespaces);
+        }
+
+        private static bool IsTestNamespace(string namespaceName)
+        {
+            return _testNamespacePattern.IsMatch(namespaceName);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext compilationStartContext)
+        {
+            InitializeApprovedNamespaces();
+            if (ApprovalFilesExist())
```

Comment:
So this analyzer will just silently do nothing if the additional files don't exist? Is that the behavior we want vs. reporting a diagnostic?

Reply:
@pakdev When I turn on this rule, ASW will be the only client that will actually have those files. For other clients, that aren't ready for this rule, should they be forced to create a ruleset where they disable this rule? Are other clients doing this already?

Reply:
I'm a little torn on that. On one hand, it is nice to make an assumption that you have at least one file that can be consumed. However, our goal was that people do not have to create a custom .ruleset to override things in general. I would lean more towards that usability workflow and say that we shouldn't complain if we're missing any files.

Reply:
The rules we enable in this package should be what we wanted enabled for _all_ C# projects at NI. If this rule _only_ applies to ASW, ASW should enable it specifically. Having it NOOP is opaque not just to projects that don't care about it but also to ASW that maybe configures something wrong.

Reply:
That is a good point Pete. Ram, is the idea that only ASW would consume this?

Reply:
@pakdev @jryckman , It will be used by ASW initially. However, there is nothing in it that prevents others from using it.

Reply:
I think the best approach is to default the rule to off and then have a specific ruleset in ASW that inherits from the NI.ruleset contained here. It can then turn on the rule and add additional files. That way, the default can be off and we still get the the error reporting behavior that Pete mentioned, which does make sense.

Reply:
sounds good.

Reply:
done.